mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp

```diff
@@ -0,0 +1,99 @@
+
+#include "mlir/Dialect/Rock/Tuning/RockTuning.h"
+
+namespace mlir {
+namespace rock {
+
+// Brute-force search in incremental order
+void createGemmTuningRangeBF(struct TunableParams *newSpace,
+                             RockGemmWrapperInterface gemmOp) {
+
+  OpBuilder b(gemmOp.getContext());
+  GemmFeatures currentFeatures = gemmOp.getGemmFeatures();
+  if (bitEnumContainsAll(currentFeatures, GemmFeatures::mfma)) {
+    // XDLOPS
+    for (uint32_t gemmMPerBlock : ValidRangeXdlopsGemmParams[0]) {
+      for (uint32_t gemmNPerBlock : ValidRangeXdlopsGemmParams[1]) {
+        for (uint32_t gemmKPerBlock : ValidRangeXdlopsGemmParams[2]) {
+          for (uint32_t gemmMPerWave : ValidRangeXdlopsGemmParams[3]) {
+            for (uint32_t gemmNPerWave : ValidRangeXdlopsGemmParams[4]) {
+              for (uint32_t gemmKPack : ValidRangeXdlopsGemmParams[5]) {
+                XdlopsGemmParamsAttr gemmParams =
+                    b.getAttr<XdlopsGemmParamsAttr>(
+                        gemmKPerBlock, gemmMPerBlock, gemmNPerBlock, gemmKPack,
+                        gemmMPerWave, gemmNPerWave);
+                newSpace->tuningRange.push_back(
+                    gemmParams.cast<RockTuningParamAttrInterface>());
+              }
+            }
+          }
+        }
+      }
+    }
+  } else {
+    // Non-XDLOPS
+    for (uint32_t blockSize : ValidRangeGeneralGemmParams[0]) {
+      for (uint32_t gemmMPerBlock : ValidRangeGeneralGemmParams[1]) {
+        for (uint32_t gemmNPerBlock : ValidRangeGeneralGemmParams[2]) {
+          for (uint32_t gemmKPerBlock : ValidRangeGeneralGemmParams[3]) {
+            for (uint32_t gemmMPerThread : ValidRangeGeneralGemmParams[4]) {
+              for (uint32_t gemmNPerThread : ValidRangeGeneralGemmParams[5]) {
+
+                GeneralGemmParamsAttr gemmParams =
+                    b.getAttr<GeneralGemmParamsAttr>(
+                        blockSize, gemmKPerBlock, gemmMPerBlock, gemmNPerBlock,
+                        /*kPerThread=*/1, gemmMPerThread, gemmNPerThread,
+                        /*kpack=*/1);
+                newSpace->tuningRange.push_back(
+                    gemmParams.cast<RockTuningParamAttrInterface>());
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  newSpace->numHeuristicQuick = 0;
+}
+
+TunableParams *createTunableParams(ModuleOp &mod) {
+  struct TunableParams *newSpace;
+  newSpace = new TunableParams();
+
+  // create range and heuristic
+  WalkResult findPrimary =
+      mod->walk([&](rock::RockGemmWrapperInterface op) -> WalkResult {
+        createGemmTuningRangeBF(newSpace, op);
+        newSpace->primaryOpType = op.getKernelType();
+        WalkResult::interrupt();
+      });
+
+  return newSpace;
+}
+
+bool tuningSetParam(ModuleOp &mod, ParamEntry *paramEntry) {
+  WalkResult setPrimary =
+      mod->walk([&](rock::RockGemmWrapperInterface op) -> WalkResult {
```

Comment:
What will happen if there are more than one op that satisfy `rock::RockGemmWrapperInterface`? Say a module with 2 convolutions; a module with 2 gemms, or a module with a convolution and a gemm